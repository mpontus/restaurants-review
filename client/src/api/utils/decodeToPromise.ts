import * as t from "io-ts";
import { failure } from "io-ts/lib/PathReporter";

/**
 * Validate input against io-ts schema and return result as a promise
 */
export const decodeToPromise = <A, O, I>(
  validator: t.Type<A, O, I>,
  input: I
): Promise<A> =>
  validator
    .decode(input)
    .fold(
      errors => Promise.reject(new Error(failure(errors).join("\n"))),
      value => Promise.resolve(value)
    );
