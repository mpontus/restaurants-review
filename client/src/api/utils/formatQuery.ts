import qs from "querystring";

/**
 * Filter undefined values from object
 */
const filterObject = (obj: Record<string, any>): object =>
  Object.keys(obj).reduce(
    (acc, key) => (obj[key] === undefined ? acc : { ...acc, [key]: obj[key] }),
    {}
  );

/**
 * Format query string from object after filtering undefined values
 */
export const formatQuery = (query: object) => qs.stringify(filterObject(query));
