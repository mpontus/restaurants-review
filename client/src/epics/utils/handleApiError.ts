import { of, OperatorFunction } from "rxjs";
import { catchError } from "rxjs/operators";
import { RequestError } from "../../models/RequestError";

/**
 * Helper function to handle API error
 *
 * Wraps error caught from source observable into RequestError and
 * passes it onto specified action creator before merging it into the
 * result stream.
 */
export const handleApiError = <T, R>(
  failureActionCreator: (e: any) => R
): OperatorFunction<T, T | R> =>
  catchError(error =>
    of(failureActionCreator(RequestError.fromApiError(error)))
  );
