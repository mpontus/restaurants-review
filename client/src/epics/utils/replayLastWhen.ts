import { merge, Observable, OperatorFunction } from "rxjs";
import { withLatestFrom } from "rxjs/operators";

/**
 * Replays the last item emitted by the source observable, whenever
 * selector emits.
 */
export const replayLastWhen = <T>(
  selector$: Observable<any>
): OperatorFunction<T, T> => source$ =>
  merge(source$, selector$.pipe(withLatestFrom(source$, (_, val) => val)));
