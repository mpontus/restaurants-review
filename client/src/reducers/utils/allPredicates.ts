/**
 * Helper type
 */
type Predicate<T, T1 extends T> = (input: T) => input is T1;

/**
 * All Predicates
 *
 * Creates a predicate which matches all of the passed predicates.
 */
export function allPredicates<T, T1 extends T>(
  predicate1: Predicate<T, T1>,
  ...rest: Array<(value: T1) => boolean>
): Predicate<T, T1>;

export function allPredicates<T, T1 extends T, T2 extends T1>(
  predicate1: Predicate<T, T1>,
  predicate2: Predicate<T1, T2>,
  ...rest: Array<(value: T2) => boolean>
): Predicate<T, T2>;

export function allPredicates<T, T1 extends T, T2 extends T1, T3 extends T2>(
  predicate1: Predicate<T, T1>,
  predicate2: Predicate<T1, T2>,
  predicate3: Predicate<T2, T3>,
  ...rest: Array<(value: T3) => boolean>
): Predicate<T, T3>;

export function allPredicates(...predicates: Array<(value: any) => boolean>) {
  return (input: any) => predicates.every(predicate => predicate(input));
}
