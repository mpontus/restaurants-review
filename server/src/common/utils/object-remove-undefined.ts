/**
 * Remove undefined values from an object
 */
export const objectRemoveUndefined = <T>(
  // tslint:disable-next-line:no-any
  obj: Partial<T> & { [K in string]?: any },
): Partial<T> =>
  Object.keys(obj).reduce(
    (acc, key) =>
      obj[key] === undefined ? acc : Object.assign(acc, { [key]: obj[key] }),
    {},
  );
