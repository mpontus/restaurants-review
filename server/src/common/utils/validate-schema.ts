// tslint:disable:promise-function-async
import * as t from 'io-ts';
import { failure } from 'io-ts/lib/PathReporter';

/**
 * Validate Schema Error
 *
 * Error dispatched when schema validation fails.
 */
export class ValidateSchemaError extends Error {}

/**
 * Validate object against the given schema
 *
 * Used to achieve type-safety on external objects, which shape can
 * not be trusted. Dispatches {@link ValidateSchemaError} when schema
 * validation fails.
 */
export const validateSchema = <A, O, I>(
  validator: t.Type<A, O, I>,
  input: I,
): Promise<A> =>
  validator
    .decode(input)
    .fold(
      errors =>
        Promise.reject(new ValidateSchemaError(failure(errors).join('\n'))),
      value => Promise.resolve(value),
    );
