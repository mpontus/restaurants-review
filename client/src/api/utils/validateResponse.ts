import { AxiosResponse } from "axios";
import * as t from "io-ts";
import { decodeToPromise } from "./decodeToPromise";

/**
 * Validate axios response against io-ts schema and cast it to
 * corresponding type.
 */
export const validateResponse = <A, O, I>(schema: t.Type<A, O, I>) => (
  response: AxiosResponse<I>
): Promise<A> => decodeToPromise(schema, response.data);
