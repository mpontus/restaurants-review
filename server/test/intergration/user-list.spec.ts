import supertest from 'supertest';
import { initApp } from '../utils/init-app';
import { resetDb } from '../utils/reset-db';

let nestApp: any;

beforeAll(async () => {
  nestApp = await initApp();
});

afterAll(() => nestApp.close());

beforeEach(() => resetDb(nestApp));

describe('List Users', () => {
  const seed = require('../seed/multiple-users');

  beforeEach(() => seed.run(nestApp));

  describe('when user is admin', () => {
    const authSeed = require('../seed/admin-user');

    beforeEach(() => authSeed.run(nestApp));

    it('should succeed', async () => {
      const response = await supertest(nestApp.getHttpServer())
        .get(`/users`)
        .set('Authorization', `Bearer ${seed.authSeed}`)
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });

    describe('when limit is provided', () => {
      it('should succeed', async () => {
        const response = await supertest(nestApp.getHttpServer())
          .get(`/users?take=2`)
          .set('Authorization', `Bearer ${seed.authSeed}`)
          .expect(200);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when offset is provided', () => {
      it('should succeed', async () => {
        const response = await supertest(nestApp.getHttpServer())
          .get(`/users?skip=3`)
          .set('Authorization', `Bearer ${seed.authSeed}`)
          .expect(200);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when limit is not a number', () => {
      it('should fail', async () => {
        const response = await supertest(nestApp.getHttpServer())
          .get(`/users?take=foo`)
          .set('Authorization', `Bearer ${seed.authSeed}`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when limit is negative', () => {
      it('should fail', async () => {
        const response = await supertest(nestApp.getHttpServer())
          .get(`/users?take=-1`)
          .set('Authorization', `Bearer ${seed.authSeed}`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when limit is too large', () => {
      it('should fail', async () => {
        const response = await supertest(nestApp.getHttpServer())
          .get(`/users?take=9000`)
          .set('Authorization', `Bearer ${seed.authSeed}`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when offset is not a number', () => {
      it('should fail', async () => {
        const response = await supertest(nestApp.getHttpServer())
          .get(`/users?skip=foo`)
          .set('Authorization', `Bearer ${seed.authSeed}`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });

    describe('when offset is negative', () => {
      it('should fail', async () => {
        const response = await supertest(nestApp.getHttpServer())
          .get(`/users?skip=-1`)
          .set('Authorization', `Bearer ${seed.authSeed}`)
          .expect(400);

        expect(response.body).toMatchSnapshot();
      });
    });
  });

  describe.skip('when user is not an admin', () => {
    const authSeed = require('../seed/regular-user');

    beforeEach(() => authSeed.run(nestApp));

    it('should fail', async () => {
      const response = await supertest(nestApp.getHttpServer())
        .get(`/users?skip=9000`)
        .set('Authorization', `Bearer ${seed.authSeed}`)
        .expect(401);

      expect(response.body).toMatchSnapshot();
    });
  });
});
