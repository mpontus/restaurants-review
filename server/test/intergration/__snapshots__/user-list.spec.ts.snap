// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Users when user is admin should succeed 1`] = `
Object {
  "items": Array [
    Object {
      "email": "ket@vavojis.ax",
      "id": "56a8809b-3fc0-5400-8bbd-512e242cc91d",
      "name": "Joshua Wood",
      "roles": Array [
        "admin",
      ],
    },
    Object {
      "email": "huhur@cverkob.cy",
      "id": "38103bba-3208-5171-b3e7-d88c8cad95de",
      "name": "Hannah Burke",
      "roles": Array [
        "user",
      ],
    },
    Object {
      "email": "iloracwur@jom.ag",
      "id": "ec96adce-52f1-5090-ab6e-c362f97336fd",
      "name": "Luis Weber",
      "roles": Array [
        "user",
      ],
    },
    Object {
      "email": "sotputu@ezev.hm",
      "id": "9d636843-0d39-58f3-be26-e500aa3c5134",
      "name": "Vincent Valdez",
      "roles": Array [
        "user",
      ],
    },
    Object {
      "email": "ewsod@diato.pa",
      "id": "649a1a68-e236-5bd4-af1d-5312127650f3",
      "name": "Timothy Willis",
      "roles": Array [
        "user",
      ],
    },
    Object {
      "email": "bavva@kihusmo.nf",
      "id": "ff4c62f5-c4f9-5984-a34b-f26d5229480b",
      "name": "Paul Jensen",
      "roles": Array [
        "user",
      ],
    },
    Object {
      "email": "taneg@nosmo.tw",
      "id": "41a5bad8-9cb4-5881-b296-2eb316df862c",
      "name": "Betty Stokes",
      "roles": Array [
        "user",
      ],
    },
    Object {
      "email": "nehkih@wamfuga.sx",
      "id": "92f3bdce-8f07-5645-96b9-b63feffe241e",
      "name": "Max Burgess",
      "roles": Array [
        "user",
      ],
    },
    Object {
      "email": "betano@jur.gt",
      "id": "92d3c50e-bfaa-5dd2-a076-14b8111e2914",
      "name": "Dennis Watts",
      "roles": Array [
        "user",
      ],
    },
    Object {
      "email": "ficfem@ti.ng",
      "id": "7859b162-1226-5b8e-82af-2f22a95fc264",
      "name": "Billy Gonzalez",
      "roles": Array [
        "user",
      ],
    },
  ],
  "total": 12,
}
`;

exports[`List Users when user is admin when limit is negative should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "min": "take must not be less than 0",
      },
      "property": "take",
      "target": Object {
        "skip": 0,
        "take": -1,
      },
      "value": -1,
    },
  ],
  "statusCode": 400,
}
`;

exports[`List Users when user is admin when limit is not a number should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "max": "take must not be greater than 50",
        "min": "take must not be less than 0",
      },
      "property": "take",
      "target": Object {
        "skip": 0,
        "take": null,
      },
      "value": null,
    },
  ],
  "statusCode": 400,
}
`;

exports[`List Users when user is admin when limit is provided should succeed 1`] = `
Object {
  "items": Array [
    Object {
      "email": "ket@vavojis.ax",
      "id": "56a8809b-3fc0-5400-8bbd-512e242cc91d",
      "name": "Joshua Wood",
      "roles": Array [
        "admin",
      ],
    },
    Object {
      "email": "huhur@cverkob.cy",
      "id": "38103bba-3208-5171-b3e7-d88c8cad95de",
      "name": "Hannah Burke",
      "roles": Array [
        "user",
      ],
    },
  ],
  "total": 12,
}
`;

exports[`List Users when user is admin when limit is too large should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "max": "take must not be greater than 50",
      },
      "property": "take",
      "target": Object {
        "skip": 0,
        "take": 9000,
      },
      "value": 9000,
    },
  ],
  "statusCode": 400,
}
`;

exports[`List Users when user is admin when offset is negative should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "min": "skip must not be less than 0",
      },
      "property": "skip",
      "target": Object {
        "skip": -1,
        "take": 10,
      },
      "value": -1,
    },
  ],
  "statusCode": 400,
}
`;

exports[`List Users when user is admin when offset is not a number should fail 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    Object {
      "children": Array [],
      "constraints": Object {
        "min": "skip must not be less than 0",
      },
      "property": "skip",
      "target": Object {
        "skip": null,
        "take": 10,
      },
      "value": null,
    },
  ],
  "statusCode": 400,
}
`;

exports[`List Users when user is admin when offset is provided should succeed 1`] = `
Object {
  "items": Array [
    Object {
      "email": "sotputu@ezev.hm",
      "id": "9d636843-0d39-58f3-be26-e500aa3c5134",
      "name": "Vincent Valdez",
      "roles": Array [
        "user",
      ],
    },
    Object {
      "email": "ewsod@diato.pa",
      "id": "649a1a68-e236-5bd4-af1d-5312127650f3",
      "name": "Timothy Willis",
      "roles": Array [
        "user",
      ],
    },
    Object {
      "email": "bavva@kihusmo.nf",
      "id": "ff4c62f5-c4f9-5984-a34b-f26d5229480b",
      "name": "Paul Jensen",
      "roles": Array [
        "user",
      ],
    },
    Object {
      "email": "taneg@nosmo.tw",
      "id": "41a5bad8-9cb4-5881-b296-2eb316df862c",
      "name": "Betty Stokes",
      "roles": Array [
        "user",
      ],
    },
    Object {
      "email": "nehkih@wamfuga.sx",
      "id": "92f3bdce-8f07-5645-96b9-b63feffe241e",
      "name": "Max Burgess",
      "roles": Array [
        "user",
      ],
    },
    Object {
      "email": "betano@jur.gt",
      "id": "92d3c50e-bfaa-5dd2-a076-14b8111e2914",
      "name": "Dennis Watts",
      "roles": Array [
        "user",
      ],
    },
    Object {
      "email": "ficfem@ti.ng",
      "id": "7859b162-1226-5b8e-82af-2f22a95fc264",
      "name": "Billy Gonzalez",
      "roles": Array [
        "user",
      ],
    },
    Object {
      "email": "ban@muhe.kw",
      "id": "7a0b3665-4ad8-5ace-b3d3-02258764cbd6",
      "name": "Emily Herrera",
      "roles": Array [
        "user",
      ],
    },
    Object {
      "email": "potsa@lo.sv",
      "id": "454da2fe-bada-5f2e-81a8-64dd43469f27",
      "name": "Ethel Bennett",
      "roles": Array [
        "user",
      ],
    },
  ],
  "total": 12,
}
`;

exports[`List Users when user is not an admin should fail 1`] = `
Object {
  "error": "Forbidden",
  "message": "Forbidden resource",
  "statusCode": 403,
}
`;
